src/geant4/app/src/DagSolidMaterial.cc

```diff
@@ -0,0 +1,139 @@
+  // one element for each nuclide, this makes making G4 materials much easier
+  std::map<int, G4Element*> g4_elements;
+  g4_elements = get_g4elements(g4_isotopes);
+
+  // from the element map, now generate G4Materials
+  std::map<std::string, G4Material*> g4_materials;
+  g4_materials = get_g4materials(g4_elements, material_library);
+
+  return g4_materials;
+}
+
+
+// Given a pyne material library, get hold of the materials and return the collection
+// of nuclides in the problem *note* using the name istope is not correct but follows
+// the Geant4 style
+std::map<int, G4Isotope*> get_g4isotopes(std::map<std::string, pyne::Material> material_library) {
+  // map to store the isotopes
+  std::map<int, G4Isotope*> g4isotopes;
+  pyne::Material sum_mat;
+
+  // parameters for new isotope
+  std::string name; // name of the isotope
+  int iz, n; // atomic number, nucleon number
+  double a; // atomic mass
+
+  // loop over the materials present and produce a single new material which contains all the isotopes
+  for (std::map<std::string, pyne::Material>::const_iterator it = material_library.begin() ; it != material_library.end() ; ++it) {
+    sum_mat = sum_mat + (it->second);
+  }
+
+  // loop through the comp map of this material
+  pyne::comp_iter mat_it;
+  // loop over the nuclides in the composition, making a new g4isotope for each one
+  for (mat_it = sum_mat.comp.begin() ; mat_it != sum_mat.comp.end() ; ++mat_it) {
+    int nuc_id = mat_it->first;
+    G4Isotope* new_iso = new G4Isotope(name = pyne::nucname::name(nuc_id), iz = pyne::nucname::znum(nuc_id), n = pyne::nucname::anum(nuc_id),
+                                       a = (pyne::atomic_mass(nuc_id) * g / mole));
+    g4isotopes[nuc_id] = new_iso;
+  }
+
+  return g4isotopes;
+}
+
+// this may seem odd to a geant4 user, but since we may care about
+// specific nuclides (g4isotopes) we need to make sure that our materials
+// have as much data regarding them as possible, since g4materials can only
+// be made of elements, we make an element for each g4isotope
+std::map<int, G4Element*> get_g4elements(std::map<int, G4Isotope*> isotope_map) {
+  std::map<int, G4Element*> element_map;
+  std::map<int, G4Isotope*>::iterator it;
+
+  std::string name, symbol; // element name and symbol
+  int ncomponents ; // number of constituent isotopes
+  double abundance; //abundance of each isotope
+  for (it = isotope_map.begin() ; it != isotope_map.end() ; ++it) {
+    G4Element* tmp = new G4Element(name = pyne::nucname::name(it->first),
+                                   symbol = pyne::nucname::name(it->first),
+                                   ncomponents = 1);
+    tmp->AddIsotope(isotope_map[it->first], abundance = 100.0 * perCent);
+    element_map[it->first] = tmp;
+  }
+
+  return element_map;
+}
+
+// using each g4element make the geant4 version of each pyne material
+std::map<std::string, G4Material*> get_g4materials(std::map<int, G4Element*> element_map,
```

Comment:
Note: I wasn't able to trace all the uses of `get_g4materials` up the call chain, but just a heads up that since this `std::map` contains pointers that are declared using `new` (on the heap), the ultimate owner of this data structure is responsible for explicitly `delete`'ing these values to avoid memory leaks.

[1] https://stackoverflow.com/a/19970551/1015951